src/lib/snarky_group_map/group_map.ml

```diff
@@ -0,0 +1,470 @@
+      else if 2*a > 3*j + 3 then acc
+      else let a' = a and b' = b + 1 in
+          go ~a:a' ~b:b' ~acc ~s ~j ~tbl
+    else F.zero
+    in
+    let a = 0 and b = 0 in
+    go ~a ~b ~acc:F.zero ~s ~j ~tbl
+
+  let iter_j s max t_powers tbl =
+    let rec go ~j ~acc ~s ~max ~t_powers ~tbl =
+      let acc' = iter_a s j tbl in
+      let acc'' = F.(+) acc (t_mul acc' j t_powers) in
+      if j < max then
+        let j' = j + 1 in
+        go ~j:j' ~acc:acc'' ~s ~max ~t_powers ~tbl
+      else acc''
+    in go ~j:0 ~acc:F.zero ~s ~max ~t_powers ~tbl
+
+  (* Xi(t) = Ni(t)/Di(t)  for i = 1, 2, 3
+   * N1(t) = A^2 . t . sum from j = 0 to j = 4 of [
+   *    sum for 2a+3b=3j of ( n1_(a,b) A^a . B^b ) . t^j ]
+   * D1(t) = sum from j = 0 to j = 5 of [
+   *    sum for 2a+3b=3j of ( d1_(a,b) A^a . B^b ) . t^j ]
+   *)
+  let make_x1 tvar =
+    let t_powers = powers tvar 15 in
+      let nt = iter_j 0 4 t_powers n1tbl in
+      let dt = iter_j 0 5 t_powers d1tbl in
+      F.(Params.a * t_mul (F.of_int 2) 1 t_powers * nt / dt)
+
+  (* N2(t) = sum from j = 0 to j = 6 of [
+   *    sum for 2a+3b=3j of ( n2_(a,b) A^a . B^b ) . t^j ]
+   * D2(t) = 144At . sum from j = 0 to j = 4 of [
+   *    sum for 2a+3b=3j of ( d2_(a,b) A^a . B^b ) . t^j ]
+   *)
+  let make_x2 tvar =
+    let t_powers = powers tvar 15 in
+      let nt = iter_j 0 6 t_powers n2tbl in
+      let dt = iter_j 0 4 t_powers d2tbl in
+      F.(nt / (t_mul (F.of_int 144 * Params.a) 1 t_powers * dt))
+
+  (* N3(t) = sum from j = 0 to j = 15 of [
+   *    sum for 2a+3b=3j+3 of ( n3_(a,b) A^a . B^b ) . t^j ]
+   * D3(t) = A . sum from j = 0 to j = 5 of [
+   *    sum for 2a+3b=3j of ( d31_(a,b) A^a . B^b ) . t^j ]
+   *    .
+   *    sum from j = 0 to j = 10 of [
+   *    sum for 2a+3b=3j of ( d32_(a,b) A^a . B^b ) . t^j ]
+   *)
+  let make_x3 tvar =
+    let t_powers = powers tvar 15 in
+      let nt = iter_j 1 15 t_powers n3tbl in
+      let dt1 = iter_j 0 5 t_powers d31tbl in
+      let dt2 = iter_j 0 10 t_powers d32tbl in
+        F.(nt / (Params.a * dt1 * dt2))
+end
+
+module Make_unchecked (F : Unchecked_field_intf)(Params : sig val a : F.t val b : F.t end) = struct
+  include Make_group_map(F)(Params)
+
+  let%test_unit "iter_a (not special case) test" =
+    Quickcheck.test ~trials:2 Int.gen ~f:(fun j ->
+        let a = iter_a 0 (j mod 4) n1tbl
+        and b = naive_iter_a 0 (j mod 4) n1tbl in
+        printf "iter_a: %s\n%!" (Sexp.to_string (F.sexp_of_t a));
+        printf "naive_iter_a: %s\n%!" (Sexp.to_string (F.sexp_of_t b));
+     assert (F.equal a b)
+    )
+
+  let%test_unit "iter_a (special case) test" =
+    Quickcheck.test ~trials:2 Int.gen ~f:(fun j ->
+        let a = iter_a 1 (j mod 4) n1tbl
+        and b = naive_iter_a 1 (j mod 4) n1tbl in
+        printf "iter_a: %s\n%!" (Sexp.to_string (F.sexp_of_t a));
+        printf "naive_iter_a: %s\n%!" (Sexp.to_string (F.sexp_of_t b));
+     assert (F.equal a b)
+    )
+
+  let to_group tvar =
+    let a = make_x1 tvar in
```

Comment:
Like in the checked case, I would do

```ocaml
let f x = F.(x * x * x + Params.a * x + Params.b) in
...
```
and then maybe something like
```ocaml
let try_decode x =
  let y2 = f x in if F.is_square y2 then Some (x, Field.sqrt y) else None
in
List.find_map [make_x1; make_x2; make_x3] ~f:(fun mk -> try_decode (mk t))
|> Option.value_exn
```
This should also be a bit more efficient since you don't compute the make_xi function unless you need to.